Refactor Tests for Codebook and IntensityTable

ambrosejcarr opened the pull request:
Summary of changes:

`Codebook` 
1. Updated to allow features that previously were filtered out to be stored in the IntensityTable. IntensityTable gains an additional data variable `passes_filters` that records whether or not each feature passes filters. 
2. Updated `per_round_max_channel_decode` to store a proper distance instead of a binary one where decoded=1, failed to decode=0. Now stores `mean_over_rounds(1 - (max_channel / sum(channels)))`
3. Side-effect: `starfish.types.Features` gains `PASSES_FILTERS='passes_filters'`

`IntensityTable`
1. Take SpotAttributes object as input, not a pd.DataFrame _called_ spot_attributes (enforces parameter checking). 
2. Side-effect: updated `SpotAttributes` to _not_ require `intensity` and `radius` as those are not always available. 
3. Side-effect: gaussian_spot_detection, local_max_peak_finder, and pixel_spot_detector all undergo minor changes to support `SpotAttributes` instead of `pd.DataFrame`

`local_max_peak_finder`
1. Squelched a numpy FutureWarning. 

`combine_adjacent_features`
1. Refactored this into a class `CombineAdjacentFeatures` and broke code into sub-methods to make more testable (output is identical). 

```diff
+ 26 files worth of tests
- original intensity_table and codebook tests. 
```

### Of note: 
All API tests pass, including tests of each data type that mimic the notebooks (ISS, DARTFISH, MERFISH). The only changes to the notebook subset tests that were necessary was to subset the intensity_tables using the new `passes_filter` fields to restrict the tests to counting features that pass filters.

ttung commented on starfish/codebook.py:
```diff
@@ -442,40 +456,45 @@ def metric_decode(
 
-        exceeds_distance = np.where(metric_outputs > max_distance)[0]
-        target_index = pd.Index(targets)
+        passes_filters = np.ones_like(targets, dtype=np.bool)
```
does it make sense to fill it in one pass?

like `passes_filters = logical_or(norms > min_intensity, metric_outputs < max_distance)`

personally, i feel this makes more sense.  i don't know if it actually works, because i suck at numpy, but I feel it should.

ttung commented on starfish/codebook.py:
```diff
@@ -442,40 +456,45 @@ def metric_decode(
 
-        return intensities
+        # norm_intensities is a DataArray, make it back into an IntensityTable
+        return IntensityTable(norm_intensities)
```
any perf implications for this?

ttung replied:
test if mutating one data set changes the other.

ambrosejcarr replied:
![image](https://user-images.githubusercontent.com/5239152/45222513-cc6d2e00-b282-11e8-8c98-84961b496730.png)

View, not copy. I think we're OK. 

I didn't add a test to the suite, since we would have to return a non-`IntensityTable` to do so.

dganguli commented on notebooks/py/DARTFISH_Pipeline_-_Human_Occipital_Cortex_-_1_FOV.py:
```diff
@@ -225,6 +225,18 @@ def compute_magnitudes(stack, norm_order=2):
 # EPY: END code
 
+# EPY: START code
+np.unique(spot_intensities['passes_filters'], return_counts=True)
```
i'm getting confused between 'filter' -- for image processing and 'filter' -- for annotating rows in an intensity table that meet certain criteria. is there a better name for filter in the latter case? e.g., passes_criteria, passes_thresholds, passes_*?

ambrosejcarr replied:
`is_valid` ?

ttung commented on starfish/codebook.py:
```diff
@@ -513,24 +532,33 @@ def _view_row_as_element(array: np.ndarray) -> np.ndarray:
-        intensities[Features.TARGET] = (
-            Features.AXIS, target_index)
+        # a code passes filters if it decodes successfully
+        passes_filters = np.ones_like(targets, dtype=np.bool)
```
two passes again.

ttung commented on starfish/intensity_table.py:
```diff
@@ -88,10 +90,12 @@ def empty_intensity_table(
 
         """
-        cls._verify_spot_attributes(spot_attributes)
+        if not isinstance(spot_attributes, SpotAttributes):
```
Unclear why we validate the data type here where we don't really elsewhere.

ambrosejcarr replied:
Added check to `from_spot_data.` 👍

ttung replied:
Actually, I think you can forgo all this verification.  That is what type annotations is supposed to do for you.  The only place we don't have type annotation coverage is the notebooks, and it's likely it'll bomb out anyway if it's the wrong type.

dganguli commented on starfish/codebook.py:
```diff
@@ -513,24 +532,33 @@ def _view_row_as_element(array: np.ndarray) -> np.ndarray:
+        # found in the non-maximal channels.
+        max_intensities = intensities.max(Indices.CH.value)
+        round_intensities = intensities.sum(Indices.CH.value)
+        distance = 1 - (max_intensities / round_intensities).mean(Indices.ROUND.value)
```
this quality score is interesting -- but not an established one. perhaps leave a TODO to further explore this? and an issue

dganguli commented on starfish/codebook.py:
```diff
@@ -513,24 +532,33 @@ def _view_row_as_element(array: np.ndarray) -> np.ndarray:
+
+        intensities[Features.TARGET] = (Features.AXIS, targets.astype('U'))
+        intensities[Features.DISTANCE] = (Features.AXIS, distance)
+        intensities[Features.PASSES_FILTERS] = (Features.AXIS, passes_filters)
```
right now, we filter (hopefully to rename to threshold, see comment above) decoded features on area and intensity. we should ALSO filter on distance/quality. Perhaps submit an Issue for this?

ttung commented on starfish/test/codebook/test_from_json.py:
```diff
@@ -0,0 +1,108 @@
+
+    # dump codebook to disk
+    codebook_data: List = codebook_json_data_factory()
+    directory = tempfile.mkdtemp()
```
https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

ttung commented on starfish/test/codebook/test_from_json.py:
```diff
@@ -0,0 +1,108 @@
+    shutil.rmtree(directory)
+
+
+@pytest.mark.skip('Tested elsewhere now that we use slicedimage backend. TODO Confirm and delete.')
```
nuke

![nuclear-explosion3](https://user-images.githubusercontent.com/280924/45178528-d03d7980-b1ca-11e8-8d60-ef8fdafc9fd7.jpg)

dganguli commented on starfish/intensity_table.py:
```diff
@@ -307,32 +298,15 @@ def from_image_stack(
         x = np.arange(xmin, xmax)
 
-        pixel_coordinates = pd.DataFrame(
+        attribute_data = pd.DataFrame(
```
rename to feature_attribute_data?

ambrosejcarr commented on starfish/intensity_table.py:
```diff
@@ -307,32 +298,15 @@ def from_image_stack(
-        image_size = cropped_data.shape[:3]
-
-        return IntensityTable.from_spot_data(intensity_data, pixel_coordinates, image_size)
+        attribute_data[Features.SPOT_RADIUS] = np.full(attribute_data.shape[0], fill_value=0.5)
```
Leave a comment about these == pixels, therefore diameter should be easily calculated from the diameter. Think about area of 1 pixel and how to calculate this.

ambrosejcarr replied:
There is a really interesting question about how we should calculate radius vs area.

ambrosejcarr replied:
Added note to #380

ambrosejcarr commented on notebooks/py/DARTFISH_Pipeline_-_Human_Occipital_Cortex_-_1_FOV.py:
```diff
@@ -225,6 +225,18 @@ def compute_magnitudes(stack, norm_order=2):
 spot_intensities, results = psd.run(zero_norm_stack)
 # EPY: END code
 
+# EPY: START code
```
Think about removing this. Are these needed?

ttung commented on starfish/test/codebook/test_from_json.py:
```diff
@@ -0,0 +1,108 @@
+    codebook = Codebook.from_code_array(codebook_array)
+
+    # Dump it to a temporary file
+    directory = tempfile.mkdtemp()
```
https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

dganguli commented on starfish/spots/_detector/combine_adjacent_features.py:
```diff
@@ -14,120 +18,358 @@ class ConnectedComponentDecodingResult(NamedTuple):
+
+        iterable = tqdm(region_properties)
+        results = mapfunc(applyfunc, iterable)
+        spot_attrs, passes_area_filter = zip(*results)
```
let's make sure that when we filter spot attribute tables, we're filtering on the same quantities across the different decoders. right now, the max_channel_per_hyb_round decoder filters on area and intensity. this decoder filters on area only. ALL decoders should filter on area, intensity, and quality/distance. this will allow for better unification of our API, pipeline, etc. Perhaps leave a more craftfully worded version of his as an Issue?

ambrosejcarr replied:
Updated #499

ttung commented on starfish/test/codebook/test_metric_decode.py:
```diff
@@ -0,0 +1,110 @@
+    Codebook with two codewords describing an experiment with three channels and two imaging rounds.
+    Both codes have two "on" channels.
+    """
+    codebook_array = [
```
does it make sense to piggy back on `codebook_array_factory`?

ambrosejcarr replied:
Yes. As you suggested, I have made a follow-up ticket to aggregate these factories into a single module: #526.

ttung replied:
Why can't I use :stately-ambrose: here? :(

ambrosejcarr commented on starfish/test/spots/detector/pixel_spot_detector/test_run.py:
```diff
@@ -0,0 +1,7 @@
+# fixtures needed: codebook, ImageStack
```
Nuke

ambrosejcarr commented:
I addressed all comments and uploaded fixed notebooks since the float PR likely won't be in until after the weekend. Previously I had waited until the floats PR to put the fixes in.

dganguli commented on starfish/codebook.py:
```diff
@@ -442,40 +456,45 @@ def metric_decode(
-        target_index.values[exceeds_distance] = 'None'
-        target_index.values[intensity_mask] = 'None'
+        # only targets with low hamming distances and high intensities should be retained
+        passes_filters = np.logical_and(
```
mind blown. didn't know about np.logical_and

dganguli commented on starfish/codebook.py:
```diff
@@ -442,40 +456,45 @@ def metric_decode(
-        # remove targets associated with distant codes
-        target_index.values[exceeds_distance] = 'None'
-        target_index.values[intensity_mask] = 'None'
+        # only targets with low hamming distances and high intensities should be retained
```
nuke word hamming, distance is generalized via the 'metric' argument.

dganguli commented on starfish/codebook.py:
```diff
@@ -442,40 +456,45 @@ def metric_decode(
-        # mask low intensity features
-        intensity_mask = np.where(norms < min_intensity)[0]
-
         metric_outputs, targets = self._approximate_nearest_code(
```
i don't remember if this was true in master or not. but we are taking a performance hit if we:

slower behavior
1. decode
2. threshold on intensity

versus.

faster behavior
1. threshold on intensity
2. decode

in the latter, we have less rows of data to pass to sklearn.nearestneighbors, which i believe is an n^2 operation.  on the other hand, the slower behavior does let us keep all distances computed for all pixels. 

my intuition  is that there should be an option in this function that lets you toggle between behavior 1 and behavior 2. perhaps this is outside the context for this PR -- but what do you think?

ambrosejcarr replied:
Spoke offline, current implementation is OK due to use of approximate nearest neighbors algorithm, which is much faster than n^2. 

However, I created an issue to address this, as @dganguli correctly points out that we could make this much faster through intelligent pre-filtering  #532.

dganguli commented on starfish/types/_constants.py:
```diff
@@ -40,3 +40,5 @@ class Features:
     CODE_VALUE = 'v'
     SPOT_RADIUS = 'radius'
     DISTANCE = 'distance'
+    PASSES_FILTERS = 'passes_filters'
```
I still think this should be called PASSES_THRESHOLDS. filters get's conflated with image filtering, which I find deeply confusing.

ambrosejcarr commented:
> Why can't I use :stately-ambrose: here? :(

`#starfish-github` translated it! Good to know...
